### PR TITLE
PRS-383 Implement syscall `sol_read_subaccount`

### PR DESCRIPTION
**Pull request by s-medvedev**

### CODE REVIEW

File: syscalls/src/subaccount.rs [hunk 1]
```diff
@@ -979,6 +979,85 @@ declare_builtin_function!(
...
+        let data = borrowed.data();
+        let start = offset as usize;
+        let end = start
+            .checked_add(length as usize)
+            .ok_or(InstructionError::InvalidArgument)?;
+        let src = data
+            .get(start..end)
+            .ok_or(InstructionError::InvalidArgument)?;
+        let dst = translate_slice_mut::<u8>(memory_mapping, buff, length, check_aligned)?;
+        dst.copy_from_slice(src);
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs:1044]: **[F-001 · medium] Unmetered byte copy.** `copy_from_slice` moves up to `length` bytes (≤ `MAX_PERMITTED_DATA_LENGTH`) but the syscall only charges `syscall_base_cost`. On the cache-hit path (`existing = Some`, ~L1022) no length-proportional cost is charged at all; on the miss path `data_len_cost` covers the load, not this copy. The `sol_memcpy` family meters per byte — recommend charging `length / cpi_bytes_per_unit` (or the mem-op rate) on both paths. Otherwise a program can `load` a large subaccount once and `read` from it in a loop, copying MBs for ~base cost each (compute-fairness / mild DoS).

File: syscalls/src/subaccount.rs [hunk 2]
```diff
@@ -979,6 +979,85 @@ declare_builtin_function!(
...
+            consume_compute_meter(invoke_context, data_len_cost)?;
+            invoke_context
+                .transaction_context
+                .add_subaccount(subaccount_pubkey, loaded)?
+        };
+
+        let borrowed = invoke_context
+            .transaction_context
+            .accounts()
+            .try_borrow_subaccount(subaccount_index)?;
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs:1033]: **[F-002 · medium, →high if persisted] No `set_subaccount_mark()`.** `load` marks the entry right after borrowing (`set_subaccount_mark()`, "so end-of-tx persistence stays in the subaccount lane"); `read` borrows immutably here and never marks. On a missing subaccount the miss path (~L1024-1035) registers `AccountSharedData::default()` (empty). Please verify end-of-tx persistence does not write an unmarked / unmutated / empty registry entry back to `subaccount_storage_address(pubkey)` (phantom-account / unintended-write risk). Suggested test: a missing-subaccount read must leave `bank.get_account(storage_addr).is_none()`.

File: syscalls/src/subaccount.rs [hunk 3]
```diff
@@ -979,6 +979,85 @@ declare_builtin_function!(
...
+                memory_mapping,
+                check_aligned,
+                invoke_context,
+                &instruction_context,
+            )?
+        };
+        compute_subaccounts_time.stop();
+        invoke_context.timings.compute_subaccounts_us += compute_subaccounts_time.as_us();
+
+        let existing = invoke_context
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs]: **[F-004 · low] Triplicated find-or-add block.** This `find_index_of_subaccount … else { get_account_shared_data + data_len_cost + add_subaccount }` sequence is now copy #3 (also in `SyscallCreateSubaccount` and `load_subaccount_impl`). Recommend extracting `fn find_or_add_subaccount(invoke_context, &pubkey) -> Result<usize>` so a metering fix (F-001) or a mark fix (F-002) is applied once, not three times.

File: syscalls/src/subaccount.rs [hunk 4]
```diff
@@ -979,6 +979,85 @@ declare_builtin_function!(
...
+    fn rust(
+        invoke_context: &mut InvokeContext,
+        seeds_addr: u64,
+        seeds_len: u64,
+        buff: u64,
+        length: u64,
+        offset: u64,
+        memory_mapping: &mut MemoryMapping,
+    ) -> Result<u64, Error> {
+        let mut load_subaccount_time = Measure::start("load_subaccount");
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs:1004]: **[F-006 · low] Read time attributed to load buckets.** The measure is named `"load_subaccount"` and accumulates into `timings.load_subaccounts_us` (~L1054-1055), so read latency reports as load latency. Consider a `read_subaccount` measure / a distinct timing field, or a comment if the shared bucket is intentional.

File: syscalls/src/lib.rs [hunk 5]
```diff
@@ -454,6 +454,7 @@ pub fn create_program_runtime_environment_v1<'a, 'ix_data>(
     result.register_function("sol_create_subaccount", SyscallCreateSubaccount::vm)?;
     result.register_function("sol_load_subaccount_rust", SyscallLoadSubaccountRust::vm)?;
     result.register_function("sol_load_subaccount_c", SyscallLoadSubaccountC::vm)?;
+    result.register_function("sol_read_subaccount", SyscallReadSubaccount::vm)?;
```
Comment (BoogVAr) [inline syscalls/src/lib.rs:457]: **[F-007 · info] Unconditional registration / no runtime feature gate.** `sol_read_subaccount` is registered unconditionally (consistent with the sibling subaccount syscalls; the allocator just below uses `register_feature_gated_function!`). Also, `load_subaccount_impl` runtime-checks `stricter_abi_and_runtime_constraints && account_data_direct_mapping`; `read` has no such guard. `read` doesn't depend on direct-mapping, so likely fine — please confirm this matches the fork's syscall-activation strategy.

File: programs/sbf/tests/programs.rs [hunk 6]
```diff
@@ -6444,3 +6444,162 @@ fn test_program_sbf_subaccount_cpi_increment() {
...
+    payload.extend_from_slice(content);
+    payload
+}
+
+/// Happy path: create a subaccount with `content`, then read the full range
+/// back through `sol_read_subaccount`. The program asserts the bytes match,
+/// so an Ok status proves the syscall returned the exact stored data.
+#[test]
+#[cfg(feature = "sbf_rust")]
+fn test_program_sbf_subaccount_read_success() {
```
Comment (BoogVAr) [inline programs/sbf/tests/programs.rs:6479]: **[F-005 · low] Coverage gaps.** The 5 tests cover full / offset / missing / out-of-range. Because `do_create=true` creates+loads+unloads first, the subsequent read hits the cache — so these are untested: (1) cross-tx read of a subaccount persisted by a *prior* tx (the non-empty miss path that charges `data_len_cost`); (2) `length == 0` (returns `Ok`, even for a missing subaccount); (3) a read of an already-loaded subaccount in one instruction (the path the F-001 metering & F-002 mark concerns live on); (4) a no-side-effect assertion after a missing read.

**Review by BoogVAr — COMMENTED** [review]
## Deep review — `sol_read_subaccount` (PRS-383)

Reviewed jointly with neonlabsorg/solana-sdk#10 (one coupled feature). Recommendation: **approved with comments** — 0 critical, 0 high, 3 medium, 3 low, 2 info.

**Verified good:** memory-safe — `translate_slice_mut(buff,length)` bounds the writable dest, `data.get(start..end)` + `checked_add` bound the source, and `copy_from_slice` operands are equal-length by construction (no panic / no OOB). PDA isolation holds — the subaccount address derives from `program_id` and the base key must be present in the instruction, so a program can only read its own subaccounts. Tests assert the real returned bytes and the exact `InstructionError::InvalidArgument`.

Inline comments below: F-001 (metering), F-002 (missing `set_subaccount_mark`), F-004 (duplicated block), F-005 (coverage), F-006 (timing), F-007 (registration).

**F-003 · medium (process):** the `parasol/solana-sdk` submodule is bumped to `c747007`, which is exactly the *head of solana-sdk#10's PR branch* — not a merged commit on `parasol-fork-dev`. If #10 is squashed/rebased on merge, this pointer orphans. Please merge solana-sdk#10 first, then repin this submodule to the post-merge SHA before merging agave#51.

**Review by BoogVAr — COMMENTED** [review]
_(Duplicate submission — please disregard; see the review immediately above for the full summary and inline comments.)_

File: syscalls/src/lib.rs (hunk 5 above)
Comment (s-medvedev) [reply syscalls/src/lib.rs:457]: Won't fix

File: programs/sbf/tests/programs.rs (hunk 6 above)
Comment (s-medvedev) [reply programs/sbf/tests/programs.rs:6479]: Fixed

File: syscalls/src/subaccount.rs (hunk 4 above)
Comment (s-medvedev) [reply syscalls/src/subaccount.rs:1004]: Commented

File: syscalls/src/subaccount.rs (hunk 3 above)
Comment (s-medvedev) [reply syscalls/src/subaccount.rs]: Fixed

File: syscalls/src/subaccount.rs (hunk 2 above)
Comment (s-medvedev) [reply syscalls/src/subaccount.rs:1033]: Added check in test

File: syscalls/src/subaccount.rs (hunk 1 above)
Comment (s-medvedev) [reply syscalls/src/subaccount.rs:1044]: Fixed

[the remark this answers is not in the document]
Comment (s-medvedev) [reply syscalls/src/subaccount.rs:1044]: Fixed

File: syscalls/src/subaccount.rs [hunk 7]
```diff
@@ -48,6 +48,38 @@ const _: () = assert!(
 
 const _: () = assert!(SUBACCOUNT_ACCOUNT_VIEW_RESERVED_SIZE % BPF_ALIGN_OF_U128 == 0);
 
+fn find_or_add_subaccount(
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs:51]: ✅ **F-004 resolved** — the triplicated find-or-add block is now a single helper shared by create/load/read. Clean.

⚠️ **Re F-002 (still relevant here).** I traced the commit path. `deconstruct_into_keyed_account_shared_data` (transaction-context `transaction_accounts.rs`) drains the **entire** subaccount lane into the persisted-accounts vec under the storage address, with **no** filter on `touched_subaccounts` or `is_subaccount()`. So every subaccount this helper adds during a committed tx is written back — including a pure-read load:
- **missing subaccount** → an empty `AccountSharedData::default()` is added; on a *successful* tx it would be written back, but the fork's zero-lamport/tombstone semantics treat the all-default account as non-existent, so nothing lands. Your new zero-length test pins exactly this. 👍
- **existing subaccount** → a read-only `sol_read_subaccount` re-stores the unchanged payload at commit (write amplification + a write-version/slot bump from a logically read-only op).

Note `touched_subaccounts` exists but is never set true — `try_borrow_mut_subaccount` doesn't call `touch()` — so the drain can't simply gate on it without dropping legitimately-created subaccounts. A proper "read doesn't persist" fix needs real per-subaccount dirty tracking, which is likely out of scope here. Flagging to confirm the redundant write on existing subaccounts is acceptable for now.

File: syscalls/src/subaccount.rs [hunk 8]
```diff
@@ -979,6 +969,78 @@ declare_builtin_function!(
...
+        dst.copy_from_slice(src);
+
+        // Charge for copying `length` bytes into guest memory.
+        let compute_cost = invoke_context.get_execution_cost();
+        let copy_cost = compute_cost.mem_op_base_cost.max(
+            length
+                .checked_div(compute_cost.cpi_bytes_per_unit)
+                .unwrap_or(u64::MAX),
+        );
+        consume_compute_meter(invoke_context, copy_cost)?;
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs:1053]: ✅ **F-001 resolved** — the copy is now metered as `mem_op_base_cost.max(length / cpi_bytes_per_unit)`, matching the agave mem_ops cost model.

Nit: the charge happens *after* `dst.copy_from_slice(src)`. The mem_ops convention is charge-before-work, so an over-budget tx doesn't perform the copy first. `length` is known before the copy, so this block can move up a few lines. Low priority.

File: syscalls/src/subaccount.rs [hunk 9]
```diff
@@ -979,6 +969,78 @@ declare_builtin_function!(
...
+    fn rust(
+        invoke_context: &mut InvokeContext,
+        seeds_addr: u64,
+        seeds_len: u64,
+        buff: u64,
+        offset: u64,
+        length: u64,
+        memory_mapping: &mut MemoryMapping,
+    ) -> Result<u64, Error> {
+        // We use `load_subaccount` measure here to capture the time spent on loading subaccount
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs:1002]: ✅ **F-006** — thanks, the comment documents the deliberate reuse of the `load_subaccount` measure / `load_subaccounts_us` bucket. Fine as-is. If read-vs-load latency ever needs separating, a dedicated `read_subaccounts_us` field would do it.

File: programs/sbf/tests/programs.rs [hunk 10]
```diff
@@ -6444,3 +6444,252 @@ fn test_program_sbf_subaccount_cpi_increment() {
...
+        bank.get_account(&storage_addr).is_none(),
+        "failed tx must not persist any subaccount state",
+    );
+}
+
+/// Edge case — no subaccount: reading zero bytes from a subaccount that was never
+/// created sees empty data, so the syscall succeeds.
+#[test]
+#[cfg(feature = "sbf_rust")]
+fn test_program_sbf_subaccount_read_zero_length_missing_subaccount() {
```
Comment (BoogVAr) [inline programs/sbf/tests/programs.rs:6611]: ✅ **F-005** — great additions. This zero-length-missing test in particular is the empirical answer to F-002: a *successful* zero-length read of a never-created subaccount persists no account (`get_account(...).is_none()`), confirming the tombstone neutralization of the empty-default write-back.

File: programs/sbf/tests/programs.rs [hunk 11]
```diff
@@ -6444,3 +6444,252 @@ fn test_program_sbf_subaccount_cpi_increment() {
...
+        logs.join("\n"),
+    );
+}
+
+/// Happy path with a non-zero offset: read a window from the middle of the
+/// stored data. The program verifies the returned bytes equal
+/// `content[offset..offset+length]`.
+#[test]
+#[cfg(feature = "sbf_rust")]
+fn test_program_sbf_subaccount_read_success_already_loaded() {
```
Comment (BoogVAr) [inline programs/sbf/tests/programs.rs:6552]: Minor follow-up to F-002: this verifies the read result on a loaded subaccount, but doesn't assert the post-tx on-chain state. If you want to lock down the redundant-write behavior (or its absence), consider asserting the storage account's data / write-version after a pure-read tx. Optional.

**Review by BoogVAr — COMMENTED** [review]
## Follow-up review — new commits `0ebdd48` (Review fixes) + `66f9dc0c` (Cargo fmt)

Re-reviewed the two new commits on top of `11ad309e`. Most prior findings are addressed — nice turnaround.

**Resolved**
- **F-001** (unmetered copy) → per-byte `copy_cost = mem_op_base_cost.max(length / cpi_bytes_per_unit)` added, matching the agave mem_ops cost model. Minor: charged *after* the copy — see inline.
- **F-004** (triplicated find-or-add) → extracted `find_or_add_subaccount`, now used by all three syscalls (create/load/read).
- **F-005** (coverage) → added cross-tx read (create in tx A with a persistence assertion, read back in tx B on the cache-miss path), an already-loaded hit-path read, the zero-length-missing edge, plus negative-persistence assertions on the failing/zero-length paths. Strong.
- **F-006** (timing) → documented the deliberate reuse of the `load_subaccount` measure / `load_subaccounts_us` bucket. OK.

**ABI parameter swap** — `sol_read_subaccount(… buff, length, offset)` → `(… buff, offset, length)`. Verified coordinated across all four surfaces: the agave builtin signature, the SBF `extern` decl + its call site, the solana-sdk `define_syscall!` (`749b5b17`), and the `read_subaccount` wrapper. Internally consistent, and there are no other callers of a brand-new syscall. No action — recording for the log.

**Still open**
- **F-003** (medium, process) — the `parasol/solana-sdk` submodule now points to `749b5b17`, which is still the **head of solana-sdk#10's PR branch**, not a merged commit on `parasol-fork-dev`. Same orphan risk as before: please merge solana-sdk#10 first, then repin this submodule to the post-merge SHA.
- **F-002** (medium → confirm) — not changed by these commits. I traced the commit path: it's benign for the missing/empty case (tombstone semantics, now pinned by your zero-length test 👍), but a pure read of an **existing** subaccount still re-persists it at commit. Details + why `touched_subaccounts` can't trivially gate it are inline on `find_or_add_subaccount`.
- **F-007** (info) — read is still registered unconditionally with no direct-mapping gate (unlike `load_subaccount_impl`). Consistent with the sibling subaccount syscalls; just confirm the fork's syscall-activation strategy.

**Recommendation: approved-with-comments.** No blockers introduced by these commits. Before merge: resolve the F-003 merge-order, and confirm the F-002 read→write-back is acceptable (or track a follow-up for real subaccount dirty-tracking).

File: transaction-context/src/transaction_accounts.rs [hunk 12]
```diff
@@ -409,9 +409,31 @@ impl TransactionAccounts {
...
+        let mut sum = self.dynamic_accounts_lamports_sum.get();
+        let shared = self.subaccount_shared_fields.borrow();
+        let touched = self.touched_subaccounts.borrow();
+        for (idx, boxed) in shared.iter().enumerate() {
+            if !touched.get(idx).copied().unwrap_or(false) {
+                // SAFETY: `lamports` is only read here; this runs after
+                // execution completes, so there are no outstanding
+                // `AccountRef`/`AccountRefMut` borrows of the subaccount lane.
+                let lamports = unsafe { (*boxed.get()).lamports };
+                sum = sum.saturating_sub(lamports as u128);
```
Comment (BoogVAr) [inline transaction-context/src/transaction_accounts.rs:433]: Info (non-blocking): `saturating_sub` will silently clamp to 0 if the touched/untouched bookkeeping ever diverges from `dynamic_accounts_lamports_sum`. Given that each `add_subaccount` adds exactly the load-time lamports and an untouched subaccount is unmodified, this can't underflow today — a `debug_assert!(sum >= lamports as u128)` (or summing only touched lamports forward instead of subtracting untouched ones) would surface accounting drift in tests without touching the release path.

File: syscalls/src/subaccount.rs [hunk 13]
```diff
@@ -441,6 +448,24 @@ fn install_subaccount_data_region(
...
+    // In-bounds direct-mapped stores hit `vm_to_host` directly and never run
+    // the `access_violation_handler`, so they can't flip the touched flag on
+    // their own (only realloc/CoW accesses route through the handler, which
+    // does call `touch`). Mark the subaccount touched here so an in-place
+    // overwrite of an existing subaccount (e.g. a writable `sol_load_subaccount`
+    // followed by a same-length write) is still persisted at tx commit. A
+    // read-only load takes the `new_readonly` branch above and is intentionally
+    // left untouched so it is not re-stored. Mirrors the "writable ⇒ persist"
+    // contract the runtime applies to main accounts.
+    if is_writable {
```
Comment (BoogVAr) [inline syscalls/src/subaccount.rs:463]: Info (non-blocking): this marks the subaccount touched on writable *mapping*, before any actual store — so a writable `sol_load_subaccount` that never writes is still re-persisted (mild write amplification). That's the conservative/correct choice and mirrors the main-account "writable ⇒ persist" contract (accounts-db dedups identical writes downstream), so no change needed; just recording the residual.

**Review by BoogVAr — COMMENTED** [review]
Re-review of the new commit `1c83ae9` ("Fix change untouched subaccounts"), on top of `0ebdd48`/`66f9dc0` from the previous pass.

This commit closes the last open medium finding from the deep review (**F-002**), and **F-003** is independently resolved. Verdict moves to **approve** — only info-level residuals remain (inline).

### F-002 (persist of an untouched / empty read-added subaccount) — resolved
The deconstruct paths now filter on the dirty flag, and I traced the full mechanism end-to-end:

- Both `deconstruct_into_keyed_account_shared_data` and the no-keys deconstruct (hence `take()`) skip any subaccount whose `touched_subaccounts[idx]` is `false`.
- Every **mutation** path sets the flag: create (`set_data_length` / `set_owner`) and funding (`set_lamports` via the system-`Transfer` CPI) go through `BorrowedAccount::touch` → `accounts.touch(index_in_transaction)`, and for a subaccount that index carries `SUBACCOUNT_MARKER`; realloc / CoW routes through the access-violation handler (`accounts.touch(payload)` with the marker payload); and a **writable** data-region mapping now eagerly touches (new branch in `install_subaccount_data_region`) to cover in-place direct-mapped stores that never reach the handler.
- A pure `sol_read_subaccount` / read-only load takes the `new_readonly` branch and never touches → it is no longer re-stored, and a never-existed subaccount is no longer resurrected as an all-default tombstone.
- `get_dynamic_accounts_lamports_sum` subtracts the lamports of untouched subaccounts so the runtime balance check (`lamports_before + sum == lamports_after`) matches what is actually persisted. This is correct: an untouched subaccount is unmodified, so its current lamports equal the load-time value previously added.

Test coverage pins it well: the SBF `test_program_sbf_subaccount_read_only_does_not_restore` contrasts a read-only subaccount against a writable-overwrite one in the same create slot and asserts via `get_account_modified_slot` (read → modified-slot unchanged; modify → modified-slot advances to the Tx2 slot), plus the two unit tests for the deconstruct filter and the balance-sum exclusion.

### F-003 (submodule pinned to PR-branch head) — resolved
The submodule now points at `c5f6253f`, which is the head of (and reachable on) solana-sdk `parasol-fork-dev` — solana-sdk#10 is merged ("PRS-383 Implement sol_read_subaccount syscall (#10)"). The merge-order / orphaned-pointer risk is gone.

### Prior findings recap
F-001 metering, F-004 helper extraction, F-005 tests — resolved in the earlier commits. F-006 (shared measure bucket) documented/acceptable. F-007 (unconditional registration, no direct-mapping gate) unchanged and consistent with sibling syscalls — confirm the activation strategy at feature-gate time.

_Reviewed by code-trace; I did not run the SBF suite locally for this commit._

**Review by BoogVAr — APPROVED** [review]
Approved
